File: adventofcode/year2022/Day03.cs
namespace adventofcode.year2022
{
    public class Day03 : Solution
    {
        private string[] _input;

        public Day03(string year, string day) : base(year, day)
        {
            _input = this.LoadInputAsLines();
        }

        private int CalculatePriority(string s)
        {
            var priority = (int) s[0];
            return priority > 96 ? priority - 96 : priority - 38;
        }

        public override string PartOne()
        {
            var total = 0;
            foreach (var sack in _input)
            {
                var sack1 = new HashSet<char>(sack.Substring(0, sack.Length / 2).ToCharArray().Distinct());
                var sack2 = new HashSet<char>(sack.Substring(sack.Length / 2).ToCharArray().Distinct());

                sack1.IntersectWith(sack2);
                total += CalculatePriority(string.Join("", sack1));
            }
            
            return Convert.ToString(total);
        }

        public override string PartTwo()
        {
            var total = 0;
            for (int i = 0; i < _input.Length; i+=3)
            {
                var sack1 = new HashSet<char>(_input[i].ToCharArray().Distinct());
                var sack2 = new HashSet<char>(_input[i + 1].ToCharArray().Distinct());
                var sack3 = new HashSet<char>(_input[i + 2].ToCharArray().Distinct());

                sack1.IntersectWith(sack2);
                sack1.IntersectWith(sack3);
                total += CalculatePriority(string.Join("", sack1));
            }
            
            return Convert.ToString(total);
        }
        
    }
}

File: adventofcode/year2020/day03.py
from __future__ import annotations
from adventofcode.common import Solution
from functools import reduce


class Day03(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)
        self._input = self._load_input_as_lines()

    def part_one(self):
        trees = x = y = 0
        while y < len(self._input) - 1:
            y += 1
            x = (x + 3) % len(self._input[y])
            trees += 1 if self._input[y][x] == '#' else 0
        return trees

    def part_two(self):
        all_trees = []
        for delta_x, delta_y in ((1, 1), (3, 1), (5, 1), (7, 1), (1, 2)):
            trees = x = y = 0
            while y < len(self._input) - 1:
                y += delta_y
                x = (x + delta_x) % len(self._input[y])
                trees += 1 if self._input[y][x] == '#' else 0
            all_trees.append(trees)
        return reduce(lambda acc, tree: acc * tree, all_trees)
    